File: MATLAB/Rx/ch_estim.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% CHANNEL ESTIMATION
% Perform channel estimation using the subcarriers of the LTF
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
function [Hinv] = ch_estim(ltf)

global ltf_fft;
global lts_len;

% v = [-26:-1,1:26]; % locations of the valid subcarriers
% Hinv = ones(lts_len,1);


% subcarriers of L1
ltf_fft1 = fftshift(fft(ltf((lts_len/2)+(1:lts_len))));
% subcarriers of L2
ltf_fft2 = fftshift(fft(ltf((lts_len/2)+lts_len+(1:lts_len))));
% average of L1 and L2
ltf_fft_avg = (ltf_fft1+ltf_fft2)/2;
% estimation
% Hinv(v+33) = ltf_fft((lts_len/2)+1+v)./(ltf_fft_avg((lts_len/2)+1+v));
Hinv = ltf_fft./ltf_fft_avg;

File: PYTHON/RX/ch_estim.py
import numpy as np
from RX import constants as c


# This function provides channel estimation based off the LTF
def ch_estim(s):
    ltf_fft = c.ltf_fft
    stf_len = c.stf_len
    lts_len = c.lts_len

    window_1_start = (lts_len // 2)  # start sample index offset by STF length plus 32
    window_1_end = window_1_start + lts_len  # start sample index offset by STF length plus LTF length plus SIG field length
    ltf = s[window_1_start:window_1_end]
    s_fft1 = np.fft.fft(ltf)
    s_fft1 = np.fft.fftshift(s_fft1)

    window_2_start = lts_len // 2 + lts_len
    window_2_end = window_2_start + lts_len  # start sample index offset by STF length plus 32 plus SIG field length plus 64
    s_fft2 = np.fft.fft(s[window_2_start:window_2_end])
    s_fft2 = np.fft.fftshift(s_fft2)

    s_fft = (s_fft1 + s_fft2) / 2

    hinv = ltf_fft / s_fft

    return hinv
